src/Juvix/Data/NumThreads.hs

```diff
@@ -0,0 +1,49 @@
+module Juvix.Data.NumThreads
+  ( NumThreads (NumThreadsAuto),
+    defaultNumThreads,
+    numThreadsOne,
+    numThreads,
+    readNumThreads,
+    mkNumThreads,
+  )
+where
+
+import GHC.Conc as GHC
+import Juvix.Extra.Strings qualified as Str
+import Juvix.Prelude.Base.Foundation
+import Prelude qualified
+
+-- | The number of jobs must be at least 1
+data NumThreads
+  = NumThreads Int
+  | NumThreadsAuto
+  deriving stock (Eq, Ord, Generic)
+
+instance Show NumThreads where
+  show = \case
+    NumThreads i -> show i
+    NumThreadsAuto -> Str.auto
+
+numThreadsOne :: NumThreads
+numThreadsOne = NumThreads 1
+
+-- | TODO improve auto with import tree
+numThreads :: (MonadIO m) => NumThreads -> m Int
+numThreads = \case
+  NumThreads i -> return i
+  NumThreadsAuto -> do
+    nc <- liftIO GHC.getNumCapabilities
+    return (max 1 (min 6 (nc - 2)))
```

Comment:
Is this based on the benchmark experiments you've done?

Reply:
It is. I have now refined it a bit based on our previous conversation. Basically, we'll use the minimum of number of processors divided by two and 8 (magic number that I've found is an ok limit). @paulcadman mentioned that there is a bug in GHC where using more cores can lead to unexpected performance loss.